vdr/types/common.go

```diff
@@ -70,3 +71,18 @@ type ResolveMetadata struct {
 	// Allow DIDs which are deactivated
 	AllowDeactivated bool
 }
+
+// DocWriterDecorator wraps a DocWriter with a function that's called before the target writer is called.
+type DocWriterDecorator struct {
+	Target    DocWriter
+	Decorator func(document did.Document, metadata DocumentMetadata) error
+}
+
+func (d DocWriterDecorator) Write(document did.Document, metadata DocumentMetadata) error {
```

Comment:
If the decorator fails, nothing is written. If the docWriter fails, connection will still be made....
The ordering and validity is very important and must be put in our specifications first.
If the decorator pattern is valid, depends on the specs.